File: problem-solving/beecrowd/1025-onde-esta-o-marmore/src/main.cpp
#include <bits/stdc++.h>

using namespace std;

int binarySearch(vector<unsigned> arr, int left, int right, int n)
{
  while (left <= right) {
    int middle = left + (right - left) / 2;

    if (arr[middle] == n and (arr[middle - 1] != n or middle == 0)) return middle;

    if (arr[middle] < n) left = middle + 1;
    else right = middle - 1;
  }

  // If we reach here, then element was not present
  return -1;
}
int main() {

  vector<vector<string>> cases;
  unsigned q, n;

  while (true)
  {
    cin >> n >> q;
    if (n == 0 and q == 0) break;

    unsigned marble, query;
    vector<unsigned> marbles;
    int index;
    vector<string> case_outputs;
    string output;

    // Raju's marbles;
    for (int i=0; i < n; i++){
      cin >> marble;
      marbles.push_back(marble);
    }

    sort(marbles.begin(), marbles.end());

    for (int i=0; i < q; i++){
      cin >> query;
      index = binarySearch(marbles, 0, marbles.size() -1 , query) + 1;

      output = (index == 0) ?
        to_string(query) + " not found\n" :
        to_string(query) + " found at " + to_string(index) + "\n";

      case_outputs.push_back(output);
    }
    cases.push_back(case_outputs);
  }

  int i = 0;
  for (vector<string> c : cases){
    i++;
    cout << "CASE# " << i << ":\n";
    for (string output : c) cout << output;
  }

  return 0;
}

File: problem-solving/beecrowd/1025-onde-esta-o-marmore/src/main.rs
use std::io;

fn main() {
    let mut cases : Vec<Vec<String>>=vec![];
    let (mut q, mut  n) : (u16, u16);

    loop{
        let (mut index, mut marble_string, mut query_string);
        let mut marble : u16;
        let mut query : u16;
        let mut marbles_vector : Vec<u16>=vec![];
        let mut case_outputs : Vec<String>=vec![];
        let mut output : String;
        let first_input = input();

        if first_input == "0 0" {break}

        let input_vector : Vec<&str>= first_input.split(" ").collect();

        n = input_vector[0].parse().unwrap();
        q = input_vector[1].parse().unwrap();

        //Raju's marbles..
        for _i in 0..n{
            marble_string = input();
            marble = marble_string.parse().unwrap();
            marbles_vector.push(marble);
        }

        //sort.
        marbles_vector.sort();


        //Meena's queries.
        for _i in 0..q {
            query_string = input();
            query = query_string.parse().unwrap();
            index = find_string(&marbles_vector, &query);

            if index == -1{
                output = format!("{} not found", query);
            }else{
                output = format!("{} found at {}", query, index);
            }

            case_outputs.push(output);
        }
        cases.push(case_outputs)
    }

    for (i, case) in cases.iter().enumerate(){
        println!("CASE# {}:", i + 1);
        for output in case.iter(){
            println!("{}", output)
        }
    }
}

fn find_string(a : &Vec<u16>, b : &u16) -> isize {
    let mut index : isize = -1;

    for (i, y) in a.iter().enumerate(){
        if b == y {
            index =  (i + 1) as isize;
            break
        }
    }
    index
}

fn input() -> String {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    input = input.trim().to_string();
    return input
}
